### PR TITLE
SALTO-5932: (Jira) Validate that the order instance reference all of it's options and validate the default values

### PR DESCRIPTION
**Pull request by SeggevHaimovich**
_Validate that the order instance reference all of it's options and validate the default values_

---

_Additional context for reviewer_

---
_Release Notes_: 
_Jira Adapter:_
* Validate that the context order instance reference all of it's options and validate the default values of the context

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_default_value.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    .filter(isInstanceChange)
+    .filter(change => getChangeData(change).elemID.typeName === FIELD_CONTEXT_TYPE_NAME)
+    .map(getChangeData)
+    .map(contextInstance => {
+      const defaultCascadingOptionIdRef = contextInstance.value.defaultValue?.cascadingOptionId
+      if (!isReferenceExpression(defaultCascadingOptionIdRef)) {
+        return undefined
+      }
+      const cascadingParent = getParent(defaultCascadingOptionIdRef.value)
+      const defaultOptionIdRef = contextInstance.value.defaultValue.optionId
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_default_value.ts:46]: I think you need:
```suggestion
      const defaultOptionIdRef = contextInstance.value.defaultValue?.optionId
```

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts [hunk 2]
```diff
@@ -6,30 +6,112 @@
...
  */
-export const fieldContextOptionsValidator: ChangeValidator = async (_changes, elementSource) => {
-  if (elementSource === undefined) {
+export const fieldContextOptionsValidator: (config: JiraConfig) => ChangeValidator = config => async changes => {
+  if (!config.fetch.splitFieldContextOptions) {
     return []
   }
-  return []
+  const orderChanges = changes
+    .filter(isAdditionOrModificationChange)
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:85]: we should add a CV that prevant removal of order without its context

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts [hunk 3]
```diff
@@ -6,30 +6,112 @@
...
+  }
+}
+
 /**
- * Verify that the context reference all the added/modified options.
+ * Verify that the orders reference all the added options, and that all orders removed from orders are removed
  */
-export const fieldContextOptionsValidator: ChangeValidator = async (_changes, elementSource) => {
-  if (elementSource === undefined) {
+export const fieldContextOptionsValidator: (config: JiraConfig) => ChangeValidator = config => async changes => {
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:80]: please add here description of the order, option and context relations (similar to what we drew in the office). If it is written somewhere else in the code then ignore this comment

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts [hunk 4]
```diff
@@ -6,30 +6,112 @@
...
+  const optionChanges = changes
+    .filter(isAdditionOrRemovalChange)
+    .filter(isInstanceChange)
+    .filter(change => getChangeData(change).elemID.typeName === FIELD_CONTEXT_OPTION_TYPE_NAME)
+
+  const addedOptionsByParent = _.groupBy(optionChanges.filter(isAdditionChange).map(getChangeData), instance =>
+    getParent(instance).elemID.getFullName(),
+  )
+  const addedOptionsErrors = _.flatMap(addedOptionsByParent, (options, parentFullName) => {
+    const order = orderByParent[parentFullName]
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:100]: please add a description

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts [hunk 5]
```diff
@@ -6,30 +6,112 @@
...
+  const afterOptions = new Set<string>(
+    collections.array
+      .makeArray(after.value.options)
+      .filter(isReferenceExpression)
+      .map(option => option.elemID.getFullName()),
+  )
+  return beforeOptions.filter(option => !afterOptions.has(option))
+}
+
+const getOrderError = (
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:51]: please split this function, one that checks for the removal error and another one that creates the error

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts [hunk 6]
```diff
@@ -6,30 +6,112 @@
...
+  orderChange: ModificationChange<InstanceElement>,
+  removedOptionsByParent: Record<string, InstanceElement[]>,
+): ChangeError | undefined => {
+  const deletedOptionsFromOrder = getDeletedOptionsFromOrderModificationChange(orderChange)
+  const removedOptions = new Set<string>(
+    (removedOptionsByParent[getParent(orderChange.data.after).elemID.getFullName()] ?? []).map(instance =>
+      instance.elemID.getFullName(),
+    ),
+  )
+  const deletedOptionsFromOrderNotInRemoved = deletedOptionsFromOrder.filter(option => !removedOptions.has(option))
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:61]: I think it is a bad name, maybe `deletedOptionsFromOrderNotInRemovedList`

File: packages/jira-adapter/src/common/fields.ts [hunk 7]
```diff
@@ -29,10 +29,23 @@ export const getContextParent = (instance: InstanceElement): InstanceElement =>
   return parent
 }
 
+export const getContextParentAsync = async (
```
Comment (shayc331) [inline packages/jira-adapter/src/common/fields.ts:32]: I think it was in the other pr, be carfull with the rebase

File: packages/jira-adapter/src/common/fields.ts [hunk 8]
```diff
@@ -29,10 +29,23 @@ export const getContextParent = (instance: InstanceElement): InstanceElement =>
...
+
 export const getContextAndFieldIds = (change: Change<InstanceElement>): { contextId: string; fieldId: string } => {
   const parent = getContextParent(getChangeData(change))
   return {
     contextId: parent.value.id,
     fieldId: getParent(parent).value.id,
   }
 }
+
+export const getOrderNameFromOption = (option: InstanceElement): string =>
```
Comment (shayc331) [inline packages/jira-adapter/src/common/fields.ts:50]: I think it was in the other pr, be carfull with the rebase

File: packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts [hunk 9]
```diff
@@ -5,116 +5,188 @@
...
+    config = _.cloneDeep(getDefaultConfig({ isDataCenter: false }))
+    config.fetch.splitFieldContextOptions = true
+    validator = fieldContextOptionsValidator(config)
+    changes = []
   })
-  it('should not return errors when changing only context', async () => {
-    changes = [toChange({ after: contextInstance })]
-    const errors = await fieldContextOptionsValidator(changes, elementsSource)
-    expect(errors).toHaveLength(0)
+  it('should return no errors when fiture flag is false', async () => {
```
Comment (shayc331) [inline packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts:57]: typo: `feature`

File: packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts [hunk 10]
```diff
@@ -5,116 +5,188 @@
...
+      expect(result).toEqual([
+        {
+          elemID: optionInstance.elemID,
+          severity: 'Error',
+          message: "This option is not being referenced by it's corresponding order",
+          detailedMessage: "The order instance context_order_child should reference all it's options",
+        },
+      ])
+    })
+    it('should return an error for cascading changes', async () => {
```
Comment (shayc331) [inline packages/jira-adapter/test/change_validators/field_contexts/field_context_options.test.ts:100]: I dont understand this test title

**Review by shayc331 — COMMENTED** [review]
looking great! add few comments

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_default_value.ts (hunk 1 above)
Comment (SeggevHaimovich) [reply packages/jira-adapter/src/change_validators/field_contexts/field_context_default_value.ts:46]: earlier we check that `contextInstance.value.defaultValue?.cascadingOptionId` is `resolvedReferenceExpression` so contextInstance.value.defaultValue must be defined

File: packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts (hunk 3 above)
Comment (SeggevHaimovich) [reply packages/jira-adapter/src/change_validators/field_contexts/field_context_options.ts:79]: Will add this in the fetch filter in the next PR.
I will add a reference here to the explanation.

File: packages/jira-adapter/src/change_validators/field_contexts/order_removal.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    removedInstances
+      .filter(
+        instance =>
+          instance.elemID.typeName === FIELD_CONTEXT_OPTION_TYPE_NAME ||
+          instance.elemID.typeName === FIELD_CONTEXT_TYPE_NAME,
+      )
+      .map(instance => instance.elemID.getFullName()),
+  )
+  const removedOrders = removedInstances.filter(instance => instance.elemID.typeName === OPTIONS_ORDER_TYPE_NAME)
+  const bla = removedOrders.filter(order => !removedOptionsAndContexts.has(getParentElemID(order).getFullName()))
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/order_removal.ts:37]: leftover?

File: packages/jira-adapter/src/change_validators/field_contexts/order_removal.ts [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  const parentName = getParentElemID(removedOrders[0]).getFullName()
+  log.error(`bla: ${bla.length}, parentName: ${parentName}`)
+  return removedOrders
+    .filter(order => !removedOptionsAndContexts.has(getParentElemID(order).getFullName()))
```
Comment (shayc331) [inline packages/jira-adapter/src/change_validators/field_contexts/order_removal.ts:41]: nit: I think it worth a function

**Review by shayc331 — APPROVED** [review]
looking great!
had small comments, approving to save a cycle

**Pull request merged**